common/src/main/java/io/pravega/common/io/serialization/RevisionDataInput.java

```diff
@@ -159,6 +178,23 @@
      */
     <K, V, M extends Map<K, V>> M readMap(ElementDeserializer<K> keyDeserializer, ElementDeserializer<V> valueDeserializer, Supplier<M> newMap) throws IOException;
 
+    /**
+     * Decodes a specific Map that has been serialized using RevisionDataOutput.writeMap() and populates the supplied 
+     * ImmutableMap builder and builds the immutable map. 
+     *
+     * This method has undefined behavior if the data starting at the current position was not encoded using RevisionDataOutput.writeMap().
+     *
+     * @param keyDeserializer   A Function that will decode a single Key of the Map from the given RevisionDataInput.
+     * @param valueDeserializer A Function that will decode a single Value of the Map from the given RevisionDataInput.
+     * @param newMapBuilder     An ImmutableMapBuilder that will create a new instance of the ImmutableMap type desired.
+     * @param <K>               Type of the Keys in the Map.
+     * @param <V>               Type of the Values in the Map.
+     * @return A new Map. If the original Map passed to RevisionDataOutput.writeMap() was null, this will return an empty map.
+     * @throws IOException If an IOException occurred.
+     */
+    <K, V> ImmutableMap<K, V> readMap(ElementDeserializer<K> keyDeserializer, ElementDeserializer<V> valueDeserializer, 
```

Comment:
Why didn't you use the same pattern as for `readCollection` above? (`C extends ...`). The `readCollection` was consistent with the other `readCollection` and `readMap`, but this `readMap` stands out as being different. Since this is a core, generic class, we should make it as flexible as possible.

Reply:
ImmutableCollection is base class for ImmutableSets, ImmutableLists, etc. 
i was trying to use the same pattern as immutableMap but couldnt find a way. 
The problem is, the builder that we give out is still typed as ImmutableCollection.Builder. 
Doing a build on it we get ImmutableCollection and not the derived type. I dont want to do an unchecked typecast just because i know the builder comes from derived type. 

I tried patterns like
```    public <T, C extends ImmutableCollection<T>> C readCollection(ElementDeserializer<T> elementDeserializer, C.Builder<T> builder) 
```

```
    public <T, B extends ImmutableCollection.Builder<T>> ImmutableCollection<T> readCollection(ElementDeserializer<T> elementDeserializer, B builder) 
```

So i decided to let the builder be populated.. the caller created the builder with correct type and when it calls build on it, the derived class' build method is invoked and we get correctly typed collection.

Reply:
Actually readMap is consistent with others in that it returns the map. 
the readCollection doesnt return a collection so that deviates from other methods..